magento/pwa-studio
# Handling invalid braintree nonce error while placing order.

## Description

This PR handles the use case of an invalid nonce being submitted to the backend during the place order mutation. This can happen if a user has created a payment nonce and has not checkout for a while, enough for the nonce to expire.

The UI would never know to refresh the payment information until the backend fails with the invalid payment nonce error when the user clicks on the place order button.

I have handled payment information errors in this PR, but also made it possible for other developers to add other error handlers in Shipping Information and  Shipping Method.

```
Checkout Page
     \__ useCheckoutPage
     \__ Shipping Information 
     \__ Shipping Method
     \__ Payment Information
              \__ usePaymentInformation
```

This PR introduces a new error class called `CheckoutError` which will be passed from the `useCheckoutPage` talon to Shipping Information, Shipping Method, and Payment Information. Each of those components should have handlers to handle this kind of an error. In this PR, I have added a handler for the `CheckoutError` in the `usePaymentInformation` talon.

## Related Issue
Closes [PWA-556](https://jira.corp.magento.com/browse/PWA-556)

### Verification Stakeholders
@jimbo 
@sirugh 

### Verification Steps
1. Go to the deployed app and add an item into the cart.
2. Go to the checkout page.
3. Complete Shipping Information, Shipping Method, and Payment Information.
4. A payment nonce should have been created and update on the cart in the backend. Now we have to invalidate that nonce to create the error. Use grapiql or any method to submit a mutation on the cart to replace the nonce with an invalid nonce.

Here is the mutation:

```
mutation setSelectedPaymentMethod($cartId: String!, $paymentNonce: String!) {
  setPaymentMethodOnCart(input: {cart_id: $cartId, payment_method: {code: "braintree", braintree: {payment_method_nonce: $paymentNonce, is_active_payment_token_enabler: false}}}) {
    cart {
      id
      selected_payment_method {
        code
        title
      }
    }
  }
}
```

and the query params:

```
{
  "cartId": "CURRENT_CART_ID",
  "paymentNonce": "ANY_STRING"
}
```

5. Once that is a success, click the place order button, it should show an error toast saying the payment method is invalid or something on those lines and the UI should ask you to re-enter the payment information.
6. Now if you click the place order button, it should work fine.

## Screenshots / Screen Captures (if appropriate)
[Demo LIink](https://adobe-my.sharepoint.com/personal/annavara_adobe_com/_layouts/15/guestaccess.aspx?guestaccesstoken=1ygavSSmPUGT49OrH8n7RL8m0z6S7ISCaDhse6LAgS0%3D&docid=2_07526f07501314c2fadb5785729a85d27&rev=1&e=sQTFBi) - The error message in the demo has been changed. I have stripped off `GraphQL error:` string from the error message. Below is the updated screen shot for that.

Error screenshot:

![image](https://user-images.githubusercontent.com/35203638/82585276-5e9e8e00-9b5b-11ea-83c5-c084819ccfef.png)


## Checklist
